Issue updating to 1.14.1: "PREV for xxxx is None"

We are trying to update from 1.13.1 to 1.14.1. However, when we run `path/to/python.exe -m conans.conan install path/to/parent/of/conanfile.py version@user/channel -g cmake -if path/to/working/dir --build missing -s "compiler=Visual Studio" -s compiler.version=15 -o tbbmalloc=True -s build_type=Release` we get an error

```
Traceback (most recent call last):
  File "build\venv\lib\site-packages\conans\model\ref.py", line 111, in loads
    _, name, version, user, channel, revision, _ = ConanFileReference.sep_pattern.split(text)
ValueError: not enough values to unpack (expected 7, got 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "build\venv\lib\site-packages\conans\client\command.py", line 377, in install
    ref = ConanFileReference.loads(args.path_or_reference)
  File "build\venv\lib\site-packages\conans\model\ref.py", line 114, in loads
    "OpenCV/1.0.6@user/stable" % text)
conans.errors.ConanException: Wrong package recipe reference C:/Users/daniel/src/flow
Write something like OpenCV/1.0.6@user/stable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "build\venv\lib\site-packages\conans\client\command.py", line 1579, in run
    method(args[0][1:])
  File "build\venv\lib\site-packages\conans\client\command.py", line 390, in install
    install_folder=args.install_folder)
  File "build\venv\lib\site-packages\conans\client\conan_api.py", line 93, in wrapper
    return f(*args, **kwargs)
  File "build\venv\lib\site-packages\conans\client\conan_api.py", line 580, in install
    no_imports=no_imports)
  File "build\venv\lib\site-packages\conans\client\manager.py", line 80, in install
    installer.install(deps_graph, keep_build)
  File "build\venv\lib\site-packages\conans\client\installer.py", line 302, in install
    self._build(nodes_by_level, keep_build, root_node, graph_info)
  File "build\venv\lib\site-packages\conans\client\installer.py", line 324, in _build
    self._handle_node_cache(node, keep_build, processed_package_refs)
  File "build\venv\lib\site-packages\conans\client\installer.py", line 404, in _handle_node_cache
    assert node.prev, "PREV for %s is None" % str(pref)
AssertionError: PREV for Boost/1.68.0-16@velo3d/thirdparty:df18e90df1fedfd9db3e0852ce38a27c76c5a54b is None

ERROR: PREV for Boost/1.68.0-16@velo3d/thirdparty:df18e90df1fedfd9db3e0852ce38a27c76c5a54b is None
```

I'm at a bit of a loss as to what is going on. 

The `version@user/channel` in the command is something we were not previously using, but we get a complaint about a missing reference if it isn't there.

The boost package  should already be on the system, if that matters.

Thanks.

We had a similar issue today on one of our mac nodes, with this error message:

```
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/model/ref.py", line 111, in loads
    _, name, version, user, channel, revision, _ = ConanFileReference.sep_pattern.split(text)
ValueError: not enough values to unpack (expected 7, got 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/command.py", line 376, in install
    ref = ConanFileReference.loads(args.path_or_reference)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/model/ref.py", line 114, in loads
    "OpenCV/1.0.6@user/stable" % text)
conans.errors.ConanException: Wrong package recipe reference pix4d-magma
Write something like OpenCV/1.0.6@user/stable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/command.py", line 1578, in run
    method(args[0][1:])
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/command.py", line 389, in install
    install_folder=args.install_folder)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/conan_api.py", line 93, in wrapper
    return f(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/conan_api.py", line 580, in install
    no_imports=no_imports)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/manager.py", line 80, in install
    installer.install(deps_graph, keep_build)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/installer.py", line 302, in install
    self._build(nodes_by_level, keep_build, root_node, graph_info)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/installer.py", line 324, in _build
    self._handle_node_cache(node, keep_build, processed_package_refs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/conans/client/installer.py", line 404, in _handle_node_cache
    assert node.prev, "PREV for %s is None" % str(pref)
AssertionError: PREV for proj/4.9.3-6@pix4d/stable:f8bda7f0751e4bc3beaa6c3b2eb02d455291c8a2 is None
```

The content of ~/.conan/data/OpenCV/3.4.2-8/pix4d/stable/metadata.json is:
```
{"recipe": {"revision": "0", "properties": {}}, "packages": {"142c4d9c725d85b48b038429a5bed8d5deeb386c": {"revision": "0", "recipe_revision": "0", "properties": {}}}}
```

and the content of ~/.conan/data/proj/4.9.3-6/pix4d/stable/metadata.json is:
```
{"recipe": {"revision": "0", "properties": {}}, "packages": {}}
```

--- a/conans/client/migrations.py
+++ b/conans/client/migrations.py
@@ -51,8 +51,25 @@ class ClientMigrator(Migrator):
         if old_version is None:
             return
 
-        if old_version < Version("1.14.0"):
-            migrate_config_install(self.cache)
+        if old_version < Version("0.25"):
+            from conans.paths import DEFAULT_PROFILE_NAME
+            default_profile_path = os.path.join(self.cache.conan_folder, PROFILES_FOLDER,
+                                                DEFAULT_PROFILE_NAME)
+            if not os.path.exists(default_profile_path):
+                self.out.warn("Migration: Moving default settings from %s file to %s"
+                              % (CONAN_CONF, DEFAULT_PROFILE_NAME))
+                conf_path = os.path.join(self.cache.conan_folder, CONAN_CONF)
+
+                migrate_to_default_profile(conf_path, default_profile_path)
+
+                self.out.warn("Migration: export_source cache new layout")
+                migrate_c_src_export_source(self.cache, self.out)
+
+        if old_version < Version("1.0"):
+            _migrate_lock_files(self.cache, self.out)
+
+        if old_version < Version("1.12.0"):
+            migrate_plugins_to_hooks(self.cache)
 
         if old_version < Version("1.13.0"):
             old_settings = """
@@ -127,25 +144,8 @@ cppstd: [None, 98, gnu98, 11, gnu11, 14, gnu14, 17, gnu17, 20, gnu20]
             # MIGRATE LOCAL CACHE TO GENERATE MISSING METADATA.json
             _migrate_create_metadata(self.cache, self.out)
 
-        if old_version < Version("1.12.0"):
-            migrate_plugins_to_hooks(self.cache)
-
-        if old_version < Version("1.0"):
-            _migrate_lock_files(self.cache, self.out)
-
-        if old_version < Version("0.25"):
-            from conans.paths import DEFAULT_PROFILE_NAME
-            default_profile_path = os.path.join(self.cache.conan_folder, PROFILES_FOLDER,
-                                                DEFAULT_PROFILE_NAME)
-            if not os.path.exists(default_profile_path):
-                self.out.warn("Migration: Moving default settings from %s file to %s"
-                              % (CONAN_CONF, DEFAULT_PROFILE_NAME))
-                conf_path = os.path.join(self.cache.conan_folder, CONAN_CONF)
-
-                migrate_to_default_profile(conf_path, default_profile_path)
-
-                self.out.warn("Migration: export_source cache new layout")
-                migrate_c_src_export_source(self.cache, self.out)
+        if old_version < Version("1.14.0"):
+            migrate_config_install(self.cache)
 
         if old_version < Version("1.14.2"):
             _migrate_full_metadata(self.cache, self.out)

--- a/conans/client/migrations.py
+++ b/conans/client/migrations.py
@@ -147,6 +147,9 @@ cppstd: [None, 98, gnu98, 11, gnu11, 14, gnu14, 17, gnu17, 20, gnu20]
                 self.out.warn("Migration: export_source cache new layout")
                 migrate_c_src_export_source(self.cache, self.out)
 
+        if old_version < Version("1.14.2"):
+            _migrate_full_metadata(self.cache, self.out)
+
 
 def _get_refs(cache):
     folders = list_folder_subdirs(cache.store, 4)
@@ -157,6 +160,50 @@ def _get_prefs(layout):
     packages_folder = layout.packages()
     folders = list_folder_subdirs(packages_folder, 1)
     return [PackageReference(layout.ref, s) for s in folders]
+
+
+def _migrate_full_metadata(cache, out):
+    # Fix for https://github.com/conan-io/conan/issues/4898
+    out.warn("Running a full revision metadata migration")
+    refs = _get_refs(cache)
+    for ref in refs:
+        try:
+            base_folder = os.path.normpath(os.path.join(cache.store, ref.dir_repr()))
+            layout = PackageCacheLayout(base_folder=base_folder, ref=ref, short_paths=None,
+                                        no_lock=True)
+            with layout.update_metadata() as metadata:
+                # Updating the RREV
+                if metadata.recipe.revision is None:
+                    out.warn("Package %s metadata had recipe revision None, migrating" % str(ref))
+                    folder = layout.export()
+                    try:
+                        manifest = FileTreeManifest.load(folder)
+                        rrev = manifest.summary_hash
+                    except Exception:
+                        rrev = DEFAULT_REVISION_V1
+                    metadata.recipe.revision = rrev
+
+                prefs = _get_prefs(layout)
+                existing_ids = [pref.id for pref in prefs]
+                for pkg_id in list(metadata.packages.keys()):
+                    if pkg_id not in existing_ids:
+                        out.warn("Package %s metadata had stalled package information %s, removing"
+                                 % (str(ref), pkg_id))
+                        del metadata.packages[pkg_id]
+                # UPDATING PREVS
+                for pref in prefs:
+                    try:
+                        pmanifest = FileTreeManifest.load(layout.package(pref))
+                        prev = pmanifest.summary_hash
+                    except Exception:
+                        prev = DEFAULT_REVISION_V1
+                    metadata.packages[pref.id].revision = prev
+                    metadata.packages[pref.id].recipe_revision = metadata.recipe.revision
+
+        except Exception as e:
+            raise ConanException("Something went wrong while migrating metadata.json files "
+                                 "in the cache, please try to fix the issue or wipe the cache: {}"
+                                 ":{}".format(ref, e))
 
 
 def _migrate_create_metadata(cache, out):
@@ -174,7 +221,7 @@ def _migrate_create_metadata(cache, out):
             try:
                 manifest = FileTreeManifest.load(folder)
                 rrev = manifest.summary_hash
-            except:
+            except Exception:
                 rrev = DEFAULT_REVISION_V1
             metadata_path = os.path.join(layout.conan(), PACKAGE_METADATA)
             if not os.path.exists(metadata_path):
@@ -186,7 +233,7 @@ def _migrate_create_metadata(cache, out):
                     try:
                         pmanifest = FileTreeManifest.load(layout.package(pref))
                         prev = pmanifest.summary_hash
-                    except:
+                    except Exception:
                         prev = DEFAULT_REVISION_V1
                     metadata.packages[pref.id].revision = prev
                     metadata.packages[pref.id].recipe_revision = metadata.recipe.revision

--- a/conans/test/functional/old/test_migrations.py
+++ b/conans/test/functional/old/test_migrations.py
@@ -1,3 +1,4 @@
+import json
 import os
 import unittest
 
@@ -7,12 +8,50 @@ from conans import __version__
 from conans.client.migrations import migrate_plugins_to_hooks, migrate_to_default_profile
 from conans.client.output import ConanOutput
 from conans.migrations import CONAN_VERSION
+from conans.model.ref import ConanFileReference
+from conans.test.utils.conanfile import TestConanFile
 from conans.test.utils.test_files import temp_folder
 from conans.test.utils.tools import TestClient
 from conans.util.files import load, save
 
 
 class TestMigrations(unittest.TestCase):
+
+    def test_migrate_revision_metadata(self):
+        # https://github.com/conan-io/conan/issues/4898
+        client = TestClient()
+        client.save({"conanfile.py": str(TestConanFile("Hello", "0.1"))})
+        client.run("create . user/testing")
+        ref = ConanFileReference.loads("Hello/0.1@user/testing")
+        layout1 = client.cache.package_layout(ref)
+        metadata = json.loads(load(layout1.package_metadata()))
+        metadata["recipe"]["revision"] = None
+        metadata["packages"]["WRONG"] = {"revision": ""}
+        metadata["packages"]["5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9"]["revision"] = None
+        metadata["packages"]["5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9"]["recipe_revision"] = None
+        save(layout1.package_metadata(), json.dumps(metadata))
+
+        client.run("create . user/stable")
+        ref2 = ConanFileReference.loads("Hello/0.1@user/stable")
+        layout2 = client.cache.package_layout(ref2)
+        metadata = json.loads(load(layout2.package_metadata()))
+        metadata["recipe"]["revision"] = "Other"
+        save(layout2.package_metadata(), json.dumps(metadata))
+
+        version_file = os.path.join(client.cache.conan_folder, CONAN_VERSION)
+        save(version_file, "1.14.1")
+        client.run("search")  # This will fire a migration
+
+        metadata_ref1 = client.cache.package_layout(ref).load_metadata()
+        self.assertEqual(metadata_ref1.recipe.revision, "2e48797069e65568befd81854aa8aaf0")
+        pkg_metadata = metadata_ref1.packages["5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9"]
+        self.assertEqual(pkg_metadata.recipe_revision, "2e48797069e65568befd81854aa8aaf0")
+        self.assertEqual(pkg_metadata.revision, "38d05b42210ca2becb43a05f9265abbe")
+        self.assertEqual(list(metadata_ref1.packages.keys()),
+                         ["5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9"])
+
+        metadata_ref2 = client.cache.package_layout(ref2).load_metadata()
+        self.assertEqual(metadata_ref2.recipe.revision, "Other")
 
     def test_migrate_config_install(self):
         client = TestClient()

